csrc/device_lower/pass/alias_memory.cpp

```diff
@@ -1385,62 +1409,350 @@ class AllocationReuseModifier : private kir::ExprMutator {
-              << alloc->buffer()->name() << " which has dtype "
-              << alloc->buffer()->dtype() << std::endl;
+      debug() << "Allocated address " << alloc->address()->toInlineString()
+              << " for T" << alloc->buffer()->name() << std::endl;
     }
+  }
 
-    current_address_ = SimplifyingIrBuilder::addExpr(address, size_bytes);
+  //! Record first reads and last writes, respecting aliased buffers
+  void recordEvents() {
+    for (auto& alloc_info : allocation_info_map_.allAllocationInfos()) {
+      if (alloc_info->alias_to) {
+        auto alias_info =
+            allocation_info_map_.getMaybeAllocationInfo(alloc_info->alias_to);
+        TORCH_CHECK(alias_info.has_value());
+        auto prev_last_read = lastAliasedRead(alias_info.value());
+        last_aliased_read_[alias_info.value()] = std::max(
+            prev_last_read, alloc_info->outer_live_interval->lastRead());
+      } else {
+        last_aliased_read_[alloc_info.get()] =
+            alloc_info->outer_live_interval->lastRead();
+      }
+    }
+
+    for (auto [alloc_info, last_read_pos] : last_aliased_read_) {
+      // Record the first write
+      auto write_it =
+          first_write_positions_
+              .emplace(alloc_info->outer_live_interval->firstWrite(), 0)
+              .first;
+      write_it->second.push_back(alloc_info);
+      // Ensure there is an entry for the last read position
+      last_read_positions_.insert(last_read_pos);
+    }
+  }
+
+  //! Pop all allocations on the top of the stack that are no longer active.
+  void reclaimMemory() {
+    if (!waiting_to_push_.empty()) {
+      // Check whether we have any allocations waiting to be pushed that are now
+      // inactive. If so, then they need to be ordered and allocated at the top
+      // of the stack before continuing.
+      if (std::any_of(
+              waiting_to_push_.begin(),
+              waiting_to_push_.end(),
+              [this](AllocationInfo* alloc_info) {
+                return lastAliasedRead(alloc_info) <= position_;
+              })) {
+        sortPushAndAssignWaiting();
+      } else {
+        // We cannot pop off the stack because there are allocations still
+        // waiting to be pushed onto the stack.
+        return;
+      }
+    }
+
+    while (!alloc_stack_.empty()) {
+      auto last_read = lastAliasedRead(alloc_stack_.back());
+      if (last_read <= position_) {
+        latest_pop_ = std::max(last_read, latest_pop_);
+        alloc_stack_.pop_back();
+      } else {
+        break;
+      }
+    }
   }
 
  private:
-  Val* current_address_ = nullptr;
+  const AllocationInfoMap& allocation_info_map_;
+
+  int position_ = -1;
+
+  // Latest position that was popped. In general, any new allocation could alias
+  // all previously popped allocations, so we use this as the most distant safe
+  // point to synchronize with the new allocation's first write (for
+  // arrive/wait).
+  int latest_pop_ = -1;
```

Comment:
Is this used?

Reply:
This was leftover from when I had planned to insert arrive/wait. We'll need it at that point but I'll remove it until then as it's not currently used.